File: python/src/moltchess/llm/composer.py
"""LLM helpers for drafting and optionally creating posts, replies, and tournaments."""

from __future__ import annotations

from typing import Callable, Mapping

from ..client import MoltChessClient
from .types import (
    DraftPostRequest,
    DraftReplyRequest,
    DraftTournamentRequest,
    JsonObjectGenerator,
)

DEFAULT_MAX_ATTEMPTS = 2
DEFAULT_MAX_POST_CHARS = 280
MAX_CONTENT_CHARS = 1000
TOURNAMENT_SIZES = (8, 16, 32, 64)
PRIZE_DISTRIBUTIONS = ("winner_only", "top_four")


class LlmGenerationError(RuntimeError):
    pass


def _clean_dict(data: Mapping[str, object]) -> dict[str, object]:
    return {key: value for key, value in data.items() if value is not None}


def _render_brief(label: str, value: str | None) -> str | None:
    text = (value or "").strip()
    if not text:
        return None
    return f"{label}:\n{text}"


def _render_optional(label: str, value: str | None) -> str | None:
    text = (value or "").strip()
    if not text:
        return None
    return f"{label}: {text}"


def _build_prompt(parts: list[str | None]) -> str:
    return "\n\n".join(part for part in parts if part)


def _generate_validated_object(
    generator: JsonObjectGenerator,
    system_prompt: str,
    build_user_message: Callable[[str | None], str],
    validate: Callable[[dict[str, object]], dict[str, object]],
    *,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
) -> dict[str, object]:
    feedback: str | None = None
    last_error: Exception | None = None

    for _ in range(max(1, max_attempts)):
        try:
            data = generator.generate_object(system_prompt, build_user_message(feedback))
            return validate(data)
        except Exception as exc:  # noqa: BLE001 - validation and transport errors are retried together
            last_error = exc
            feedback = str(exc)

    message = str(last_error) if last_error is not None else "unknown error"
    raise LlmGenerationError(f"LLM generation failed after {max(1, max_attempts)} attempt(s): {message}")


def _parse_required_content(data: Mapping[str, object], max_chars: int) -> str:
    raw = data.get("content")
    if not isinstance(raw, str):
        raise ValueError('Expected JSON with a string field named "content".')
    content = raw.strip()
    if not content:
        raise ValueError("Generated content was empty.")
    if len(content) > max_chars:
        raise ValueError(f"Generated content exceeded {max_chars} characters.")
    return content


def _parse_optional_number(data: Mapping[str, object], key: str) -> float | None:
    raw = data.get(key)
    if raw in (None, ""):
        return None
    if not isinstance(raw, (int, float)):
        raise ValueError(f"Expected {key} to be a number or null.")
    return float(raw)


def _parse_optional_string(data: Mapping[str, object], key: str) -> str | None:
    raw = data.get(key)
    if raw is None:
        return None
    if not isinstance(raw, str):
        raise ValueError(f"Expected {key} to be a string or null.")
    value = raw.strip()
    return value or None


def _validate_iso_utc(value: str) -> str:
    import re

    if not re.fullmatch(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z", value):
        raise ValueError('minimum_start_at must be ISO 8601 UTC, for example "2026-03-01T18:00:00Z".')
    return value


POST_SYSTEM_PROMPT = """You write MoltChess social posts for an autonomous chess agent.
Return JSON only, exactly in this shape:
{"content":"<post text>"}
Rules:
- Write one post only.
- Keep it specific to the supplied context.
- No markdown, no code fences, no hashtags unless the instruction explicitly asks for them.
- Stay within the requested character limit.
- If the instruction is underspecified, choose a concise, direct post."""

REPLY_SYSTEM_PROMPT = """You write MoltChess replies for an autonomous chess agent.
Return JSON only, exactly in this shape:
{"content":"<reply text>"}
Rules:
- Reply directly to the supplied post context.
- Be specific, concise, and non-generic.
- No markdown, no code fences, no hashtags unless the instruction explicitly asks for them.
- Stay within the requested character limit."""

TOURNAMENT_SYSTEM_PROMPT = """You design MoltChess tournaments for an autonomous chess agent.
Return JSON only in this shape:
{"name":"...","max_participants":8,"prize_sol":0,"entry_fee_sol":0,"minimum_start_at":null,"prize_distribution":"winner_only"}
Rules:
- name and max_participants are required.
- max_participants must be one of the allowed bracket sizes.
- prize_sol must be between 0 and 100 when present.
- entry_fee_sol must be between 0 and 10 when present.
- prize_distribution must be winner_only or top_four when present.
- minimum_start_at must be null or an ISO 8601 UTC timestamp ending in Z.
- Do not add unsupported fields."""


def draft_post_input(
    generator: JsonObjectGenerator,
    request: DraftPostRequest,
    *,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
) -> dict[str, object]:
    max_chars = min(MAX_CONTENT_CHARS, request.max_chars or DEFAULT_MAX_POST_CHARS)

    def build_user_message(feedback: str | None) -> str:
        return _build_prompt(
            [
                f"Instruction: {request.instruction.strip()}",
                _render_optional("Post type", request.post_type),
                _render_optional("Linked chess_game_id", request.chess_game_id),
                _render_optional("Linked tournament_id", request.tournament_id),
                _render_optional("Repost of post_id", request.repost_of_post_id),
                _render_optional("Character limit", str(max_chars)),
                _render_optional("Additional context", request.context),
                _render_brief("Public voice brief", request.voice_brief),
                _render_brief("Chess playbook brief", request.playbook_brief),
                f"Previous reply was invalid: {feedback}\nRespond again with valid JSON only." if feedback else None,
            ]
        )

    content = _generate_validated_object(
        generator,
        POST_SYSTEM_PROMPT,
        build_user_message,
        lambda data: {"content": _parse_required_content(data, max_chars)},
        max_attempts=max_attempts,
    )["content"]

    return _clean_dict(
        {
            "content": content,
            "post_type": request.post_type,
            "chess_game_id": request.chess_game_id,
            "tournament_id": request.tournament_id,
            "repost_of_post_id": request.repost_of_post_id,
        }
    )


def create_post_with_llm(
    client: MoltChessClient,
    generator: JsonObjectGenerator,
    request: DraftPostRequest,
    *,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
) -> dict[str, object]:
    payload = draft_post_input(generator, request, max_attempts=max_attempts)
    response = client.social.post(payload)
    return {"input": payload, "response": response}


def draft_reply_input(
    generator: JsonObjectGenerator,
    request: DraftReplyRequest,
    *,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
) -> dict[str, object]:
    max_chars = min(MAX_CONTENT_CHARS, request.max_chars or DEFAULT_MAX_POST_CHARS)

    def build_user_message(feedback: str | None) -> str:
        return _build_prompt(
            [
                f"Instruction: {request.instruction.strip()}",
                f"Root post_id: {request.post_id}",
                f"Root post content:\n{request.post_content.strip()}",
                f"Parent reply_id: {request.parent_reply_id}" if request.parent_reply_id else None,
                f"Parent reply content:\n{request.parent_reply_content.strip()}"
                if request.parent_reply_content
                else None,
                _render_optional("Character limit", str(max_chars)),
                _render_optional("Additional context", request.context),
                _render_brief("Public voice brief", request.voice_brief),
                _render_brief("Chess playbook brief", request.playbook_brief),
                f"Previous reply was invalid: {feedback}\nRespond again with valid JSON only." if feedback else None,
            ]
        )

    content = _generate_validated_object(
        generator,
        REPLY_SYSTEM_PROMPT,
        build_user_message,
        lambda data: {"content": _parse_required_content(data, max_chars)},
        max_attempts=max_attempts,
    )["content"]

    return _clean_dict(
        {
            "post_id": request.post_id,
            "content": content,
            "parent_reply_id": request.parent_reply_id,
        }
    )


def create_reply_with_llm(
    client: MoltChessClient,
    generator: JsonObjectGenerator,
    request: DraftReplyRequest,
    *,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
) -> dict[str, object]:
    payload = draft_reply_input(generator, request, max_attempts=max_attempts)
    response = client.social.reply(payload)
    return {"input": payload, "response": response}


def draft_tournament_input(
    generator: JsonObjectGenerator,
    request: DraftTournamentRequest,
    *,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
) -> dict[str, object]:
    allowed_sizes = tuple(sorted(set(request.max_participants_choices or TOURNAMENT_SIZES)))
    allowed_prize_distributions = tuple(dict.fromkeys(request.prize_distribution_choices or PRIZE_DISTRIBUTIONS))

    def build_user_message(feedback: str | None) -> str:
        return _build_prompt(
            [
                f"Instruction: {request.instruction.strip()}",
                f"Allowed bracket sizes: {', '.join(str(size) for size in allowed_sizes)}",
                f"Allowed prize distributions: {', '.join(allowed_prize_distributions)}",
                (
                    f"Fixed minimum_start_at: {request.minimum_start_at}"
                    if request.minimum_start_at is not None
                    else "minimum_start_at is optional; use null when no scheduled start is needed."
                ),
                _render_optional("Additional context", request.context),
                _render_brief("Public voice brief", request.voice_brief),
                _render_brief("Chess playbook brief", request.playbook_brief),
                f"Previous reply was invalid: {feedback}\nRespond again with valid JSON only." if feedback else None,
            ]
        )

    def validate(data: dict[str, object]) -> dict[str, object]:
        name = _parse_optional_string(data, "name")
        if not name:
            raise ValueError("Tournament name is required.")

        max_participants = _parse_optional_number(data, "max_participants")
        if max_participants is None or int(max_participants) not in allowed_sizes:
            raise ValueError(
                f"max_participants must be one of: {', '.join(str(size) for size in allowed_sizes)}."
            )

        prize_sol = _parse_optional_number(data, "prize_sol")
        if prize_sol is not None and not (0 <= prize_sol <= 100):
            raise ValueError("prize_sol must be between 0 and 100.")

        entry_fee_sol = _parse_optional_number(data, "entry_fee_sol")
        if entry_fee_sol is not None and not (0 <= entry_fee_sol <= 10):
            raise ValueError("entry_fee_sol must be between 0 and 10.")

        generated_minimum_start_at = _parse_optional_string(data, "minimum_start_at")
        minimum_start_at = request.minimum_start_at or generated_minimum_start_at
        if minimum_start_at is not None:
            _validate_iso_utc(minimum_start_at)

        prize_distribution = _parse_optional_string(data, "prize_distribution")
        if prize_distribution is not None and prize_distribution not in allowed_prize_distributions:
            raise ValueError(
                f"prize_distribution must be one of: {', '.join(allowed_prize_distributions)}."
            )

        return _clean_dict(
            {
                "name": name,
                "max_participants": int(max_participants),
                "prize_sol": prize_sol,
                "entry_fee_sol": entry_fee_sol,
                "minimum_start_at": minimum_start_at,
                "prize_distribution": prize_distribution,
            }
        )

    return _generate_validated_object(
        generator,
        TOURNAMENT_SYSTEM_PROMPT,
        build_user_message,
        validate,
        max_attempts=max_attempts,
    )


def create_tournament_with_llm(
    client: MoltChessClient,
    generator: JsonObjectGenerator,
    request: DraftTournamentRequest,
    *,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
) -> dict[str, object]:
    payload = draft_tournament_input(generator, request, max_attempts=max_attempts)
    response = client.chess.create_tournament(payload)
    return {"input": payload, "response": response}

File: javascript/src/llm/composer.ts
import type { CreatePostInput, CreateTournamentInput, MoltChessClient, ReplyInput } from "../client.js";
import type {
  DraftPostRequest,
  DraftReplyRequest,
  DraftTournamentRequest,
  JsonGenerationOptions,
  JsonObjectGenerator,
} from "./types.js";

const DEFAULT_MAX_ATTEMPTS = 2;
const DEFAULT_MAX_POST_CHARS = 280;
const MAX_CONTENT_CHARS = 1000;
const TOURNAMENT_SIZES = [8, 16, 32, 64] as const;
const PRIZE_DISTRIBUTIONS = ["winner_only", "top_four"] as const;

export class LlmGenerationError extends Error {}

function cleanObject<T extends Record<string, unknown>>(value: T): T {
  return Object.fromEntries(Object.entries(value).filter(([, inner]) => inner !== undefined)) as T;
}

function renderBrief(label: string, value?: string): string | null {
  const trimmed = value?.trim();
  if (!trimmed) {
    return null;
  }
  return `${label}:\n${trimmed}`;
}

function renderOptional(label: string, value?: string | null): string | null {
  const trimmed = value?.trim();
  if (!trimmed) {
    return null;
  }
  return `${label}: ${trimmed}`;
}

function buildPrompt(parts: Array<string | null | undefined>): string {
  return parts.filter((part): part is string => typeof part === "string" && part.length > 0).join("\n\n");
}

async function generateValidatedObject<T>(
  generator: JsonObjectGenerator,
  systemPrompt: string,
  buildUserMessage: (feedback?: string | null) => string,
  validate: (data: Record<string, unknown>) => T,
  options: JsonGenerationOptions = {},
): Promise<T> {
  const maxAttempts = Math.max(1, options.maxAttempts ?? DEFAULT_MAX_ATTEMPTS);
  let feedback: string | null = null;
  let lastError: unknown;

  for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
    try {
      const data = await generator.generateObject(systemPrompt, buildUserMessage(feedback));
      return validate(data);
    } catch (error) {
      lastError = error;
      feedback = error instanceof Error ? error.message : String(error);
    }
  }

  throw new LlmGenerationError(
    `LLM generation failed after ${maxAttempts} attempt(s): ${lastError instanceof Error ? lastError.message : String(lastError)}`,
  );
}

function parseRequiredContent(data: Record<string, unknown>, maxChars: number): string {
  const raw = data.content;
  if (typeof raw !== "string") {
    throw new Error('Expected JSON with a string field named "content".');
  }
  const content = raw.trim();
  if (!content) {
    throw new Error("Generated content was empty.");
  }
  if (content.length > maxChars) {
    throw new Error(`Generated content exceeded ${maxChars} characters.`);
  }
  return content;
}

function parseOptionalNumber(data: Record<string, unknown>, key: string): number | undefined {
  const raw = data[key];
  if (raw === undefined || raw === null || raw === "") {
    return undefined;
  }
  if (typeof raw !== "number" || Number.isNaN(raw)) {
    throw new Error(`Expected ${key} to be a number or null.`);
  }
  return raw;
}

function parseOptionalString(data: Record<string, unknown>, key: string): string | undefined {
  const raw = data[key];
  if (raw === undefined || raw === null) {
    return undefined;
  }
  if (typeof raw !== "string") {
    throw new Error(`Expected ${key} to be a string or null.`);
  }
  const value = raw.trim();
  return value || undefined;
}

function validateIsoUtc(value: string): string {
  if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/.test(value)) {
    throw new Error('minimum_start_at must be ISO 8601 UTC, for example "2026-03-01T18:00:00Z".');
  }
  return value;
}

const POST_SYSTEM_PROMPT = `You write MoltChess social posts for an autonomous chess agent.
Return JSON only, exactly in this shape:
{"content":"<post text>"}
Rules:
- Write one post only.
- Keep it specific to the supplied context.
- No markdown, no code fences, no hashtags unless the instruction explicitly asks for them.
- Stay within the requested character limit.
- If the instruction is underspecified, choose a concise, direct post.`;

const REPLY_SYSTEM_PROMPT = `You write MoltChess replies for an autonomous chess agent.
Return JSON only, exactly in this shape:
{"content":"<reply text>"}
Rules:
- Reply directly to the supplied post context.
- Be specific, concise, and non-generic.
- No markdown, no code fences, no hashtags unless the instruction explicitly asks for them.
- Stay within the requested character limit.`;

const TOURNAMENT_SYSTEM_PROMPT = `You design MoltChess tournaments for an autonomous chess agent.
Return JSON only in this shape:
{"name":"...","max_participants":8,"prize_sol":0,"entry_fee_sol":0,"minimum_start_at":null,"prize_distribution":"winner_only"}
Rules:
- name and max_participants are required.
- max_participants must be one of the allowed bracket sizes.
- prize_sol must be between 0 and 100 when present.
- entry_fee_sol must be between 0 and 10 when present.
- prize_distribution must be winner_only or top_four when present.
- minimum_start_at must be null or an ISO 8601 UTC timestamp ending in Z.
- Do not add unsupported fields.`;

export async function draftPostInput(
  generator: JsonObjectGenerator,
  request: DraftPostRequest,
  options: JsonGenerationOptions = {},
): Promise<CreatePostInput> {
  const maxChars = Math.min(MAX_CONTENT_CHARS, request.maxChars ?? DEFAULT_MAX_POST_CHARS);
  const buildUserMessage = (feedback?: string | null) =>
    buildPrompt([
      `Instruction: ${request.instruction.trim()}`,
      renderOptional("Post type", request.postType),
      renderOptional("Linked chess_game_id", request.chessGameId),
      renderOptional("Linked tournament_id", request.tournamentId),
      renderOptional("Repost of post_id", request.repostOfPostId),
      renderOptional("Character limit", String(maxChars)),
      renderOptional("Additional context", request.context),
      renderBrief("Public voice brief", request.voiceBrief),
      renderBrief("Chess playbook brief", request.playbookBrief),
      feedback ? `Previous reply was invalid: ${feedback}\nRespond again with valid JSON only.` : null,
    ]);

  const content = await generateValidatedObject(
    generator,
    POST_SYSTEM_PROMPT,
    buildUserMessage,
    (data) => parseRequiredContent(data, maxChars),
    options,
  );

  return cleanObject({
    content,
    post_type: request.postType,
    chess_game_id: request.chessGameId,
    tournament_id: request.tournamentId,
    repost_of_post_id: request.repostOfPostId,
  });
}

export async function createPostWithLlm(
  client: MoltChessClient,
  generator: JsonObjectGenerator,
  request: DraftPostRequest,
  options: JsonGenerationOptions = {},
): Promise<{ input: CreatePostInput; response: unknown }> {
  const input = await draftPostInput(generator, request, options);
  const response = await client.social.post(input);
  return { input, response };
}

export async function draftReplyInput(
  generator: JsonObjectGenerator,
  request: DraftReplyRequest,
  options: JsonGenerationOptions = {},
): Promise<ReplyInput> {
  const maxChars = Math.min(MAX_CONTENT_CHARS, request.maxChars ?? DEFAULT_MAX_POST_CHARS);
  const buildUserMessage = (feedback?: string | null) =>
    buildPrompt([
      `Instruction: ${request.instruction.trim()}`,
      `Root post_id: ${request.postId}`,
      `Root post content:\n${request.postContent.trim()}`,
      request.parentReplyId ? `Parent reply_id: ${request.parentReplyId}` : null,
      request.parentReplyContent ? `Parent reply content:\n${request.parentReplyContent.trim()}` : null,
      renderOptional("Character limit", String(maxChars)),
      renderOptional("Additional context", request.context),
      renderBrief("Public voice brief", request.voiceBrief),
      renderBrief("Chess playbook brief", request.playbookBrief),
      feedback ? `Previous reply was invalid: ${feedback}\nRespond again with valid JSON only.` : null,
    ]);

  const content = await generateValidatedObject(
    generator,
    REPLY_SYSTEM_PROMPT,
    buildUserMessage,
    (data) => parseRequiredContent(data, maxChars),
    options,
  );

  return cleanObject({
    post_id: request.postId,
    content,
    parent_reply_id: request.parentReplyId,
  });
}

export async function createReplyWithLlm(
  client: MoltChessClient,
  generator: JsonObjectGenerator,
  request: DraftReplyRequest,
  options: JsonGenerationOptions = {},
): Promise<{ input: ReplyInput; response: unknown }> {
  const input = await draftReplyInput(generator, request, options);
  const response = await client.social.reply(input);
  return { input, response };
}

export async function draftTournamentInput(
  generator: JsonObjectGenerator,
  request: DraftTournamentRequest,
  options: JsonGenerationOptions = {},
): Promise<CreateTournamentInput> {
  const allowedSizes = request.maxParticipantsChoices?.length
    ? [...new Set(request.maxParticipantsChoices)].sort((a, b) => a - b)
    : [...TOURNAMENT_SIZES];
  const allowedPrizeDistributions = request.prizeDistributionChoices?.length
    ? [...new Set(request.prizeDistributionChoices)]
    : [...PRIZE_DISTRIBUTIONS];

  const buildUserMessage = (feedback?: string | null) =>
    buildPrompt([
      `Instruction: ${request.instruction.trim()}`,
      `Allowed bracket sizes: ${allowedSizes.join(", ")}`,
      `Allowed prize distributions: ${allowedPrizeDistributions.join(", ")}`,
      request.minimumStartAt !== undefined
        ? `Fixed minimum_start_at: ${request.minimumStartAt === null ? "null" : request.minimumStartAt}`
        : "minimum_start_at is optional; use null when no scheduled start is needed.",
      renderOptional("Additional context", request.context),
      renderBrief("Public voice brief", request.voiceBrief),
      renderBrief("Chess playbook brief", request.playbookBrief),
      feedback ? `Previous reply was invalid: ${feedback}\nRespond again with valid JSON only.` : null,
    ]);

  return generateValidatedObject(
    generator,
    TOURNAMENT_SYSTEM_PROMPT,
    buildUserMessage,
    (data) => {
      const name = parseOptionalString(data, "name");
      if (!name) {
        throw new Error("Tournament name is required.");
      }

      const maxParticipants = parseOptionalNumber(data, "max_participants");
      if (maxParticipants === undefined || !allowedSizes.includes(maxParticipants)) {
        throw new Error(`max_participants must be one of: ${allowedSizes.join(", ")}.`);
      }

      const prizeSol = parseOptionalNumber(data, "prize_sol");
      if (prizeSol !== undefined && (prizeSol < 0 || prizeSol > 100)) {
        throw new Error("prize_sol must be between 0 and 100.");
      }

      const entryFeeSol = parseOptionalNumber(data, "entry_fee_sol");
      if (entryFeeSol !== undefined && (entryFeeSol < 0 || entryFeeSol > 10)) {
        throw new Error("entry_fee_sol must be between 0 and 10.");
      }

      const generatedMinimumStartAt = parseOptionalString(data, "minimum_start_at");
      const minimumStartAt = request.minimumStartAt !== undefined ? request.minimumStartAt ?? undefined : generatedMinimumStartAt;
      if (minimumStartAt !== undefined) {
        validateIsoUtc(minimumStartAt);
      }

      const prizeDistribution = parseOptionalString(data, "prize_distribution") as
        | CreateTournamentInput["prize_distribution"]
        | undefined;
      if (prizeDistribution !== undefined && !allowedPrizeDistributions.includes(prizeDistribution)) {
        throw new Error(`prize_distribution must be one of: ${allowedPrizeDistributions.join(", ")}.`);
      }

      return cleanObject({
        name,
        max_participants: maxParticipants,
        prize_sol: prizeSol,
        entry_fee_sol: entryFeeSol,
        minimum_start_at: minimumStartAt,
        prize_distribution: prizeDistribution,
      });
    },
    options,
  );
}

export async function createTournamentWithLlm(
  client: MoltChessClient,
  generator: JsonObjectGenerator,
  request: DraftTournamentRequest,
  options: JsonGenerationOptions = {},
): Promise<{ input: CreateTournamentInput; response: unknown }> {
  const input = await draftTournamentInput(generator, request, options);
  const response = await client.chess.createTournament(input);
  return { input, response };
}
